relayer/relayer/src/handlers/azero.rs

```diff
@@ -0,0 +1,249 @@
+    AzeroContract(#[from] AzeroContractError),
+
+    #[error("Eth contract error")]
+    EthContractTx(#[from] ContractError<SignedEthConnection>),
+
+    #[error("Tx was not present in any block or mempool after the maximum number of retries")]
+    TxNotPresentInBlockOrMempool,
+
+    #[error("Contract reverted")]
+    EthContractReverted,
+}
+
+pub struct AlephZeroEventHandler;
+
+impl AlephZeroEventHandler {
+    pub async fn handle_event(
+        event: ContractEvent,
+        config: Arc<Config>,
+        eth_signed_connection: Arc<SignedEthConnection>,
+    ) -> Result<(), AlephZeroEventHandlerError> {
+        let Config {
+            eth_contract_address,
+            eth_tx_min_confirmations,
+            eth_tx_submission_retries,
+            ..
+        } = &*config;
+
+        if let Some(name) = &event.name {
+            if name.eq("CrosschainTransferRequest") {
+                let data = event.data;
+
+                // decode event data
+                let CrosschainTransferRequestData {
+                    committee_id,
+                    dest_token_address,
+                    amount,
+                    dest_receiver_address,
+                    request_nonce,
+                } = get_request_event_data(&data)?;
+
+                info!(
+                    "Decoded event data: [dest_token_address: 0x{}, amount: {amount}, dest_receiver_address: 0x{}, request_nonce: {request_nonce}]",
+                    hex::encode(dest_token_address),
+                    hex::encode(dest_receiver_address)
+                );
+
+                // NOTE: for some reason, ethers-rs's `encode_packed` does not properly encode the data
+                // (it does not pad uint to 32 bytes, but uses the actual number of bytes required to store the value)
+                // so we use `abi::encode` instead (it only differs for signed and dynamic size types, which we don't use here)
+                let bytes = abi::encode(&[
+                    Token::Uint(committee_id.into()),
+                    Token::FixedBytes(dest_token_address.to_vec()),
+                    Token::Uint(amount.into()),
+                    Token::FixedBytes(dest_receiver_address.to_vec()),
+                    Token::Uint(request_nonce.into()),
+                ]);
+
+                debug!("ABI event encoding: 0x{}", hex::encode(bytes.clone()));
+
+                let request_hash = keccak256(bytes);
+                let request_hash_hex = hex::encode(request_hash);
+
+                info!("hashed event encoding: 0x{}", request_hash_hex);
+
+                let address = eth_contract_address.parse::<Address>()?;
+                let contract = Most::new(address, eth_signed_connection.clone());
+
+                while contract
+                    .needs_signature(
+                        request_hash,
+                        eth_signed_connection.address(),
+                        committee_id.into(),
+                    )
```

Comment:
Similarly as for the other half. It matters quite a bit which block is this checked against.

I would do the following:
loop {
    - check at most recent block, if `needs_signature`, then submit tx,
    - check at finalized block, if doesn't need signature, then break
}

Reply:
If we always check only agains the most recent block, then this is not correct, as most recent block can revert